src/defaultCredentials.ts

```diff
@@ -0,0 +1,85 @@
+import * as fs from "fs";
+import * as path from "path";
+
+import * as api from "./api";
+import { configstore } from "./configstore";
+import * as logger from "./logger";
+
+// Interface for a valid JSON refresh token credential, so the
+// fields must be snake_case not camelCase.
+interface RefreshTokenCredential {
+  client_id: string; // eslint-disable-line
+  client_secret: string; // eslint-disable-line
+  refresh_token: string; // eslint-disable-line
+  type: string; // eslint-disable-line
+}
+
+/**
+ * Get a path to the application default credentials JSON file.
+ */
+export async function getCredentialPathAsync(): Promise<string | undefined> {
+  const filePath = credFilePath();
+  if (!filePath) {
+    logger.debug("defaultcredentials: could not create path to default credentials file.");
+    return undefined;
+  }
+
+  const cred = getCredential();
+  if (!cred) {
+    logger.debug("defaultcredentials: no credential available.");
+    return undefined;
+  }
+
+  // We could use fs.writeFileSync() here but it's important that the caller understands
+  // that this is a somewhat expensive operation so we make it a Promise.
+  return new Promise((res, rej) => {
```

Comment:
Wouldn't this overwrite whatever is currently in that path? Is that the behavior we want?

Reply:
Yes that's the behavior we want.  I'm willing to pay this small extra cost to avoid the whole class of bugs that come with potentially out of date credentials.